### PR TITLE
Add RandomAccessFile digest methods

### PR DESCRIPTION
**Pull request by behrangsa**
Added two new methods:
    
* `digest(final MessageDigest messageDigest, final RandomAccessFile data)`
* `updateDigest(final MessageDigest digest, final RandomAccessFile data)`
    
For computing digest of `RandomAccessFile`s.
    
Performance-wise there's almost no improvements offered by these new NIO based methods, but being non-blocking they could potentially use fewer resources and provide better throughput in highly multi-threaded scenarios.

### CODE REVIEW

**Comment by behrangsa** [comment]
Not sure why the coveralls is reporting that the new methods are not tested. Tests have been added for both of them.

File: src/test/java/org/apache/commons/codec/digest/MessageDigestAlgorithmsTest.java [hunk 1]
```diff
@@ -146,6 +151,22 @@ public void testDigestFile() throws IOException {
         Assert.assertArrayEquals(digestTestData(), DigestUtils.digest(DigestUtils.getDigest(messageDigestAlgorithm),getTestFile()));
     }
 
+    @Test
+    public void testNonBlockingDigestRandomAccessFile() throws IOException {
+        Assume.assumeTrue(DigestUtils.isAvailable(messageDigestAlgorithm));
+        Assert.assertArrayEquals(digestTestData(),
+                DigestUtils.nonblockingDigest(DigestUtils.getDigest(messageDigestAlgorithm), getTestRandomAccessFile()));
+        Assert.assertArrayEquals(digestTestData(), DigestUtils.digest(DigestUtils.getDigest(messageDigestAlgorithm),getTestFile()));
```
Comment (aherbert) [inline src/test/java/org/apache/commons/codec/digest/MessageDigestAlgorithmsTest.java]: This should call `nonblockingDigest` with `getTestRandomAccessFile()`

File: src/test/java/org/apache/commons/codec/digest/MessageDigestAlgorithmsTest.java [hunk 2]
```diff
@@ -146,6 +151,22 @@ public void testDigestFile() throws IOException {
...
+    }
+
+    @Test
+    public void testNonBlockingDigestFile() throws IOException {
+        Assume.assumeTrue(DigestUtils.isAvailable(messageDigestAlgorithm));
+        Assert.assertArrayEquals(digestTestData(),
+                DigestUtils.nonblockingDigest(DigestUtils.getDigest(messageDigestAlgorithm), getTestFile()));
+        Assert.assertArrayEquals(digestTestData(), DigestUtils.digest(DigestUtils.getDigest(messageDigestAlgorithm),getTestFile()));
```
Comment (aherbert) [inline src/test/java/org/apache/commons/codec/digest/MessageDigestAlgorithmsTest.java]: This should call `nonblockingDigest`.

**Comment by aherbert** [comment]
I am also not sure why coveralls dropped. Looking at the Travis log for JDK 8 it runs the tests:
```
[INFO] Running org.apache.commons.codec.digest.MessageDigestAlgorithmsTest
[WARNING] Tests run: 88, Failures: 0, Errors: 0, Skipped: 32, Time elapsed: 4.816 s - in org.apache.commons.codec.digest.MessageDigestAlgorithmsTest
```

When I run locally `mvn clean test -Dtest=MessageDigestAlgorithmsTest jacoco:report` I get the same numbers of tests run (those skipped are for algorithms in later JDKs) and the local jacoco report has coverage of the new methods.

If you fix the new tests to make a duplicate call to the new NIO methods and push it will run again and we can see if this is some strange glitch in coveralls.

**Comment by behrangsa** [comment]
Silly me. Thanks for the review.

**Comment by behrangsa** [comment]
@aherbert I fixed those issues, but for some reason coveralls reports that they are not covered yet:

```java
    @Test
    public void testNonBlockingDigestRandomAccessFile() throws IOException {
        Assume.assumeTrue(DigestUtils.isAvailable(messageDigestAlgorithm));

        final byte[] expected = digestTestData();

        Assert.assertArrayEquals(expected,
                DigestUtils.nonblockingDigest(
                        DigestUtils.getDigest(messageDigestAlgorithm), getTestRandomAccessFile()
                )
        );

        Assert.assertArrayEquals(expected,
                DigestUtils.nonblockingDigest(
                        DigestUtils.getDigest(messageDigestAlgorithm), getTestRandomAccessFile()
                )
        );
    }

    @Test
    public void testNonBlockingDigestFile() throws IOException {
        Assume.assumeTrue(DigestUtils.isAvailable(messageDigestAlgorithm));

        final byte[] expected = digestTestData();

        Assert.assertArrayEquals(
                expected,
                DigestUtils.nonblockingDigest(
                        DigestUtils.getDigest(messageDigestAlgorithm), getTestFile()
                )
        );

        Assert.assertArrayEquals(
                expected,
                DigestUtils.nonblockingDigest(
                        DigestUtils.getDigest(messageDigestAlgorithm), getTestFile()
                )
        );
    }
```

**Comment by aherbert** [comment]
The failure is due to the post travis action to build the coveralls report. This command:
```
mvn clean cobertura:cobertura coveralls:report -Ptravis-cobertura
```
Fails to run the test:
```
[INFO] Running org.apache.commons.codec.digest.MessageDigestAlgorithmsTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.001 s <<< FAILURE! - in org.apache.commons.codec.digest.MessageDigestAlgorithmsTest
[ERROR] org.apache.commons.codec.digest.MessageDigestAlgorithmsTest  Time elapsed: 0.001 s  <<< ERROR!
java.lang.ClassCastException: [I cannot be cast to java.lang.String
	at org.apache.commons.codec.digest.MessageDigestAlgorithmsTest.checkValues(MessageDigestAlgorithmsTest.java:71)
```
This is a problem on master that needs fixing. It is not related to your PR.

**Comment by aherbert** [comment]
I fixed master to work with cobertura. If you rebase you should get coveralls reporting.

**Comment by garydgregory** [comment]
I am fine with adding more methods to handing more inputs but I am -1 to putting implementation details like "NIO" in and "nonblocking" in method names.

File: src/main/java/org/apache/commons/codec/digest/DigestUtils.java [hunk 3]
```diff
@@ -99,6 +101,33 @@
...
+    /**
+     * Reads through a File using non-blocking-io (NIO) and returns the digest for the data
+     *
+     * @param messageDigest The MessageDigest to use (e.g. MD5)
+     * @param data          Data to digest
+     * @return the digest
+     * @throws IOException On error reading from the stream
+     * @since 1.14
+     */
+    public static byte[] nonblockingDigest(final MessageDigest messageDigest, final File data) throws IOException {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/codec/digest/DigestUtils.java]: Don't use implementation details in method names, so -1 to the "nonblocking" prefix.

File: src/main/java/org/apache/commons/codec/digest/DigestUtils.java [hunk 4]
```diff
@@ -1207,6 +1236,43 @@ public static MessageDigest updateDigest(final MessageDigest digest, final File
...
+     * Reads through a RandomAccessFile and updates the digest for the data
+     * using non-blocking-io (NIO)
+     *
+     * @param digest The MessageDigest to use (e.g. MD5)
+     * @param data   Data to digest
+     * @return the digest
+     * @throws IOException On error reading from the stream
+     * @since 1.14
+     */
+    public static MessageDigest updateDigestNio(final MessageDigest digest, final RandomAccessFile data) throws IOException {
```
Comment (garydgregory) [inline src/main/java/org/apache/commons/codec/digest/DigestUtils.java]: Don't use implementation details in method names, so -1 to the "Nio" postfix.

File: src/main/java/org/apache/commons/codec/digest/DigestUtils.java (hunk 3 above)
Comment (behrangsa) [reply src/main/java/org/apache/commons/codec/digest/DigestUtils.java]: If I remove the prefix from:

```java
nonblockingDigest(final MessageDigest messageDigest, final File data)
```

will clash with an existing method name.

File: src/main/java/org/apache/commons/codec/digest/DigestUtils.java (hunk 4 above)
Comment (behrangsa) [reply src/main/java/org/apache/commons/codec/digest/DigestUtils.java]: I've added the prefix here for consistency with the other overload that takes a `File`.

**Comment by garydgregory** [comment]
Let's step back and ask: Why would we want two implementations for `File` objects? If NIO does the job better, then let's change the implementation to do that.

**Comment by behrangsa** [comment]
> Let's step back and ask: Why would we want two implementations for `File` objects? If NIO does the job better, then let's change the implementation to do that.

I thought using NIO would improve performance but in my limited tests I didn't notice a meaningful difference. It would be great if some NIO experts could chime in.

However we can:

* Get rid of `nonblockingDigest(final MessageDigest messageDigest, final File data)`
* Rename `nonblockingDigest(final MessageDigest messageDigest, final RandomAccessFile data)` to `digest(final MessageDigest messageDigest, final RandomAcessFile data)`

as the first method calls wraps the File in a RandomAccessFile and passes it to the 2nd method anyway.

File: src/main/java/org/apache/commons/codec/digest/DigestUtils.java [hunk 5]
```diff
@@ -99,6 +101,20 @@
         return updateDigest(messageDigest, data).digest();
     }
 
+    /**
+     * Reads through a RandomAccessFile using non-blocking-io (NIO) and returns the digest for the data
```
Comment (aherbert) [inline src/main/java/org/apache/commons/codec/digest/DigestUtils.java]: First line of javadoc should end with a period.

File: src/main/java/org/apache/commons/codec/digest/DigestUtils.java [hunk 6]
```diff
@@ -1207,6 +1223,29 @@ public static MessageDigest updateDigest(final MessageDigest digest, final File
         }
     }
 
+    /**
+     * Reads through a RandomAccessFile and updates the digest for the data
+     * using non-blocking-io (NIO)
```
Comment (aherbert) [inline src/main/java/org/apache/commons/codec/digest/DigestUtils.java]: First line of javadoc should end with a period.

File: src/test/java/org/apache/commons/codec/digest/DigestUtilsTest.java [hunk 7]
```diff
@@ -63,23 +67,37 @@ File getTestFile() {
...
             testFile.deleteOnExit();
         }
+
+        if (!testRandomAccessFile.delete()) {
+            testRandomAccessFile.deleteOnExit();
+        }
     }
 
-    @Test(expected=IllegalArgumentException.class)
+    @Test(expected = IllegalArgumentException.class)
```
Comment (aherbert) [inline src/test/java/org/apache/commons/codec/digest/DigestUtilsTest.java]: You have reformatted whitespace for the rest of this class.

It makes it hard to see that you have not actually changed anything below this line other than whitespace. Can you unformat please.

File: src/test/java/org/apache/commons/codec/digest/MessageDigestAlgorithmsTest.java [hunk 8]
```diff
@@ -151,6 +156,25 @@ public void testDigestFile() throws IOException {
         Assert.assertArrayEquals(digestTestData(), DigestUtils.digest(DigestUtils.getDigest(messageDigestAlgorithm),getTestFile()));
     }
 
+    @Test
+    public void testNonBlockingDigestRandomAccessFile() throws IOException {
+        Assume.assumeTrue(DigestUtils.isAvailable(messageDigestAlgorithm));
+
+        final byte[] expected = digestTestData();
+
+        Assert.assertArrayEquals(expected,
```
Comment (aherbert) [inline src/test/java/org/apache/commons/codec/digest/MessageDigestAlgorithmsTest.java]: In keeping with the rest of the class can you reformat the assertions to a single line. The 5 lines per assertion are very explicit.

File: src/test/java/org/apache/commons/codec/digest/DigestUtilsTest.java (hunk 7 above)
Comment (behrangsa) [reply src/test/java/org/apache/commons/codec/digest/DigestUtilsTest.java]: It is possible to ignore white-space changes in GitHub reviews: https://github.blog/2018-05-01-ignore-white-space-in-code-review/

File: src/main/java/org/apache/commons/codec/digest/DigestUtils.java (hunk 5 above)
Comment (behrangsa) [reply src/main/java/org/apache/commons/codec/digest/DigestUtils.java]: Fixed.

File: src/test/java/org/apache/commons/codec/digest/DigestUtilsTest.java (hunk 7 above)
Comment (behrangsa) [reply src/test/java/org/apache/commons/codec/digest/DigestUtilsTest.java]: Fixed.

File: src/test/java/org/apache/commons/codec/digest/MessageDigestAlgorithmsTest.java (hunk 8 above)
Comment (behrangsa) [reply src/test/java/org/apache/commons/codec/digest/MessageDigestAlgorithmsTest.java]: Fixed.

**Comment by garydgregory** [comment]
xref https://issues.apache.org/jira/browse/CODEC-272

**Pull request closed without merging**